omd_tyxml/omd_tyxml.ml

```diff
@@ -0,0 +1,140 @@
+open Tyxml
+
+(* TODO Document *)
+(* TODO Fix tests *)
+(* TODO self-review and cleanup *)
+
+exception Invalid_markdown of string
+
+exception Unsupported_attribute of string
+
+let of_omd_attributes attrs =
+  List.map (fun (a, v) -> Html.Unsafe.string_attrib a v) attrs
+
+(* TODO move into Omd module *)
+let rec inline_to_plaintext : Omd.inline -> string =
+  fun il ->
+  match il.il_desc with
+  (* FIXME nontail rec *)
+  | Concat xs -> List.map inline_to_plaintext xs |> String.concat ""
+  | Text s -> s
+  | Emph t -> inline_to_plaintext t
+  | Strong t -> inline_to_plaintext t
+  | Code s -> s
+  | Hard_break -> ""
+  | Soft_break -> ""
+  | Html s -> s
+  | Link l | Image l -> inline_to_plaintext l.label
+
+let of_code attrs c = Html.[code ~a:attrs [txt c]]
+
+let rec of_inline ({il_attributes; il_desc} : Omd.inline) =
+  let attrs = of_omd_attributes il_attributes in
+  match il_desc with
+  | Code c     -> of_code attrs c
+  | Concat ls  -> List.concat_map of_inline ls
+  | Emph e     -> Html.[em ~a:[] (of_inline e)]
+  | Strong s   -> Html.[strong ~a:[] (of_inline s)]
+  | Hard_break -> Html.[br ~a:[] ()]
+  (* TODO Add option for verified html ?*)
+  | Html raw   -> Html.Unsafe.[data raw]
```

Comment:
This is "correct" as long as the HTML is textual. As soon as you start using the non-textual Tyxml backends (for instance, Tyxml_js`), this is broken and you have to do more work.

Long term, I think it would be beneficial to provide a functor over the HTML module to adapt to different Tyxml instantiations (and provide a pre-applied instance for text!). The "right" solution is then to decode the HTML with lambdasoup and build actual tyxml trees. Then you can build DOM trees from markdown directly, without going through text.

Reply:
I'm not sure whether we want this generalization as part of the current PR, or as a followup, but I agree we should target it in the long run.

I think we first need to settle settle whether or not we can replace the bespoke html generator with Tyxml.